ocaml/typing/subst.ml

```diff
@@ -506,215 +502,136 @@ let rename_bound_idents scoping s sg =
           rest
-    | SigL_module(id, pres, md, rs, vis) :: rest ->
+    | Sig_module(id, pres, md, rs, vis) :: rest ->
         let id' = rename id in
         rename_bound_idents
           (add_module id (Pident id') s)
-          (SigL_module (id', pres, md, rs, vis) :: sg)
+          (Sig_module (id', pres, md, rs, vis) :: sg)
           rest
-    | SigL_modtype(id, mtd, vis) :: rest ->
+    | Sig_modtype(id, mtd, vis) :: rest ->
         let id' = rename id in
         rename_bound_idents
-          (add_modtype id (Mty_ident(Pident id')) s)
-          (SigL_modtype(id', mtd, vis) :: sg)
+          (add_modtype id (Types.Mty_ident(Pident id')) s)
+          (Sig_modtype(id', mtd, vis) :: sg)
           rest
-    | SigL_class(id, cd, rs, vis) :: rest ->
+    | Sig_class(id, cd, rs, vis) :: rest ->
         (* cheat and pretend they are types cf. PR#6650 *)
         let id' = rename id in
         rename_bound_idents
           (add_type id (Pident id') s)
-          (SigL_class(id', cd, rs, vis) :: sg)
+          (Sig_class(id', cd, rs, vis) :: sg)
           rest
-    | SigL_class_type(id, ctd, rs, vis) :: rest ->
+    | Sig_class_type(id, ctd, rs, vis) :: rest ->
         (* cheat and pretend they are types cf. PR#6650 *)
         let id' = rename id in
         rename_bound_idents
           (add_type id (Pident id') s)
-          (SigL_class_type(id', ctd, rs, vis) :: sg)
+          (Sig_class_type(id', ctd, rs, vis) :: sg)
           rest
-    | SigL_value(id, vd, vis) :: rest ->
+    | Sig_value(id, vd, vis) :: rest ->
         (* scope doesn't matter for value identifiers. *)
         let id' = Ident.rename id in
-        rename_bound_idents s (SigL_value(id', vd, vis) :: sg) rest
-    | SigL_typext(id, ec, es, vis) :: rest ->
+        rename_bound_idents s (Sig_value(id', vd, vis) :: sg) rest
+    | Sig_typext(id, ec, es, vis) :: rest ->
         let id' = rename id in
-        rename_bound_idents s (SigL_typext(id',ec,es,vis) :: sg) rest
+        rename_bound_idents s (Sig_typext(id',ec,es,vis) :: sg) rest
   in
   rename_bound_idents s [] sg
 
-let rec lazy_module_decl md =
-  { mdl_type = lazy_modtype md.md_type;
-    mdl_attributes = md.md_attributes;
-    mdl_loc = md.md_loc;
-    mdl_uid = md.md_uid }
-
-and subst_lazy_module_decl scoping s md =
-  let mdl_type = subst_lazy_modtype scoping s md.mdl_type in
-  { mdl_type;
-    mdl_attributes = attrs s md.mdl_attributes;
-    mdl_loc = loc s md.mdl_loc;
-    mdl_uid = md.mdl_uid }
-
-and force_module_decl md =
-  let md_type = force_modtype md.mdl_type in
-  { md_type;
-    md_attributes = md.mdl_attributes;
-    md_loc = md.mdl_loc;
-    md_uid = md.mdl_uid }
+module To_lazy = Types.Map_pods(Types)(Lazy_types)
 
-and lazy_functor_parameter = function
-| Types.Unit -> Unit
-| Types.Named (id, mty) -> Named (id, lazy_modtype mty)
+let to_lazy =
+  let map_signature m sg =
+    lazy (List.map (To_lazy.signature_item m) sg) |> Pod.of_lazy
```

Comment:
I was indeed worried about this. However, this would only happen if we convert a strict signature to a lazy one and then back without substituting in between. I've convinced myself that this is never the case or at least, not enough to matter in any measurable way - performance and allocations seem unchanged with this PR. In fact, I expected this to happen when we deserialise a (strict) signature and then look it up, going via a lazy one but it seems we always freshen in that case. Even if it didn't this specific case wouldn't matter since we'll start deserialising lazily soon.